### ISSUE TITLE
Transient "Error unknown features: repositories" when creating project with features set

### ISSUE THREAD

**Issue by rob-zz** [microsoft/terraform-provider-azuredevops#595]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and Azure DevOps Provider) Version

% terraform -v
Terraform v1.2.0
on darwin_arm64
+ provider registry.terraform.io/hashicorp/azuread v2.22.0
+ provider registry.terraform.io/microsoft/azuredevops v0.2.1

### Affected Resource(s)

azuredevops_project

* `azuredevops_XXXXX`

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

```
% terraform apply
╷
│ Error: unknown feature: repositories
│ 
│   with azuredevops_project.projects["testrobbert"],
│   on devops_projects.tf line 10, in resource "azuredevops_project" "projects":
│   10:   features = {
│   11:     "boards"       = "disabled"
│   12:     "repositories" = "disabled"
│   13:     "pipelines"    = "enabled"
│   14:     "testplans"    = "disabled"
│   15:     "artifacts"    = "disabled"
│   16:   }
│ 
╵
```

### Expected Behavior

Applied the features

### Actual Behavior

It returned the error above.

### Steps to Reproduce

Can't reproduce, subsequent run of 'terraform apply' succeeded.

### References

* #0000

**Comment by xuzhang3** [comment]
@robbert-nlo  Any other errors? This is  a validation error that happened during local configuration validation,  this is hard coded validation to ensure the features names are right.

**Comment by rob-zz** [comment]
@xuzhang3  No, no other errors. I know it is a validation error, but apparently the validation is flaky. I hit it again with a 'terraform plan', and again a subsequent run made it go away. Does not feel very stable:

```
~/git/devopstest/nlo-common % terraform plan        
azuread_application.app["test2robbert-prd"]: Refreshing state... [id=]
azuread_application.app["test2robbert-stg"]: Refreshing state... [id=]
azuread_application.app["testrobbert-prd"]: Refreshing state... [id=]
azuread_service_principal.sp["test2robbert-prd"]: Refreshing state... [id=]
azuread_service_principal.sp["test2robbert-stg"]: Refreshing state... [id=]
azuread_service_principal.sp["testrobbert-prd"]: Refreshing state... [id=]
azuredevops_project.projects["test2robbert"]: Refreshing state... [id=]
azuredevops_project.projects["testrobbert"]: Refreshing state... [id=]
azuredevops_environment.envs["test2robbert-stg"]: Refreshing state... [id=]
azuredevops_environment.envs["testrobbert-prd"]: Refreshing state... [id=]
azuredevops_environment.envs["test2robbert-prd"]: Refreshing state... [id=]
╷
│ Error: unknown feature: artifacts
│ 
│   with azuredevops_project_features.project-features["test2robbert"],
│   on devops_projects.tf line 23, in resource "azuredevops_project_features" "project-features":
│   23:   features = {
│   24:     "boards"       = "disabled"
│   25:     "repositories" = "disabled"
│   26:     "pipelines"    = "enabled"
│   27:     "testplans"    = "disabled"
│   28:     "artifacts"    = "disabled"
│   29:   }
│ 
╵
~/git/devopstest/nlo-common % terraform plan
azuread_application.app["testrobbert-prd"]: Refreshing state... [id=]
azuread_application.app["test2robbert-stg"]: Refreshing state... [id=]
azuread_application.app["test2robbert-prd"]: Refreshing state... [id=]
azuread_service_principal.sp["test2robbert-prd"]: Refreshing state... [id=]
azuread_service_principal.sp["test2robbert-stg"]: Refreshing state... [id=]
azuread_service_principal.sp["testrobbert-prd"]: Refreshing state... [id=]
azuredevops_project.projects["testrobbert"]: Refreshing state... [id=]
azuredevops_project.projects["test2robbert"]: Refreshing state... [id=]
azuredevops_environment.envs["testrobbert-prd"]: Refreshing state... [id=]
azuredevops_environment.envs["test2robbert-prd"]: Refreshing state... [id=]
azuredevops_environment.envs["test2robbert-stg"]: Refreshing state... [id=]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply" which may have affected this plan:

  # azuredevops_project.projects["test2robbert"] has changed
  ~ resource "azuredevops_project" "projects" {
      + features            = {}
        id                  = ""
        name                = "test2robbert"
        # (5 unchanged attributes hidden)
    }

  # azuredevops_project.projects["testrobbert"] has changed
  ~ resource "azuredevops_project" "projects" {
      + features            = {}
        id                  = ""
        name                = "testrobbert"
        # (5 unchanged attributes hidden)
    }


Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # azuredevops_project_features.project-features["test2robbert"] will be created
  + resource "azuredevops_project_features" "project-features" {
      + features   = {
          + "artifacts"    = "disabled"
          + "boards"       = "disabled"
          + "pipelines"    = "enabled"
          + "repositories" = "disabled"
          + "testplans"    = "disabled"
        }
      + id         = (known after apply)
      + project_id = ""
    }

  # azuredevops_project_features.project-features["testrobbert"] will be created
  + resource "azuredevops_project_features" "project-features" {
      + features   = {
          + "artifacts"    = "disabled"
          + "boards"       = "disabled"
          + "pipelines"    = "enabled"
          + "repositories" = "disabled"
          + "testplans"    = "disabled"
        }
      + id         = (known after apply)
      + project_id = ""
    }

Plan: 2 to add, 0 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

**Comment by xuzhang3** [comment]
This is  strange. Based on the error log,  looks like the feature names for validation has been changed.

**Comment by xuzhang3** [comment]
@robbert-nlo  #593 has the same error as yours and has a more detailed error logs. From the log,this is a concurrent  modification error.

**Comment by rob-zz** [comment]
@xuzhang3 : I'm submitter for both bugs you marked as duplicate. Not sure why they are duplicate, one is a clear crash of the plugin (which IMO should never happen) and the other seems a validation error. That said, the underlying cause could of course be the same.

Also, I'm not entirely sure what you mean by 'concurrent modification error', but if you mean that these resources were modified by multiple people at the same time, then I am surprised. This was a local terraform repository with test resources that I was working on alone at the time of the issues. No one else was working on these resources at the same time.

**Comment by xuzhang3** [comment]
@robbert-nlo Error message `fatal error: concurrent map writes`  in #593 direct this to concurrent update error.  I can reproduce this error in my local dev env. 
For  `│ Error: unknown feature: artifacts`,  potential concurrent update/read will happen at https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/service/core/resource_project_features.go#L113 when multiple goroutines call this function.
Potential  scenario for #595
1. goroutine A call function `getProjectFeatureNameMapReverse()` and update `projectFeatureNameMapReverse`, but only `boards` added to the map, other four types not.
2. At the same time, goroutine B call this function too and found that `projectFeatureNameMapReverse` is not empty (`if len(projectFeatureNameMapReverse) <= 0`), so gorountine B will use `projectFeatureNameMapReverse` directly but cannot find `artifacts` in `projectFeatureNameMapReverse`,  validation failed

**Comment by rob-zz** [comment]
@xuzhang3 Ah, I understand, you mean it is a concurrency issue internal to the extension.

**Comment by xuzhang3** [comment]
> @xuzhang3 Ah, I understand, you mean it is a concurrency issue internal to the extension.

yes, this is a provider internal bug.

**Comment by rob-zz** [comment]
Already fixed in 0.2. by #596 .
